src/AggregateFunctions/QuantileInterpolatedWeighted.h

```diff
@@ -0,0 +1,304 @@
+            auto weight = pair.getMapped();
+            value_weight_pairs.push_back(std::make_pair(value, weight));
+        }
+
+        ::sort(value_weight_pairs.begin(), value_weight_pairs.end(), [](const Pair & a, const Pair & b) { return a.first < b.first; });
+
+        Float64 accumulated = 0;
+
+        /// vector for populating and storing the cumulative sum using the provided weights.
+        /// example: [0,1,2,3,4,5] -> [0,1,3,6,10,15]
+        std::vector<Float64> weights_cum_sum;
+
+        bool first = true;
+        for (size_t idx = 0; idx < size; ++idx)
+        {
+            accumulated += value_weight_pairs[idx].second;
+
+            if (first)
+            {
+                weights_cum_sum.emplace_back(value_weight_pairs[idx].second);
+                first = false;
+            }
+            else
+            {
+                weights_cum_sum.emplace_back(accumulated);
+            }
+        }
+
+        /// The following estimation of quantile is general and the idea is:
+        /// https://en.wikipedia.org/wiki/Percentile#The_weighted_percentile_method
+
+        /// calculates a simple cumulative distribution based on weights
+        /// Note: if sum_weight is 0 then subsequent division is inf, so set the value of the exp to 0.
+        for (size_t idx = 0; idx < size; ++idx)
+            value_weight_pairs[idx].second
+                = sum_weight == 0 ? 0 : (weights_cum_sum[idx] - 0.5 * value_weight_pairs[idx].second) / sum_weight;
+
+        size_t level_index = 0;
+
+        while (level_index < num_levels)
+        {
+            /// perform linear interpolation for every level
+            auto level = levels[indices[level_index]];
+
+            size_t idx = 0;
+            if (size >= 2)
+            {
+                if (level >= value_weight_pairs[size - 2].second)
+                {
+                    idx = size - 2;
+                }
+                else
+                {
+                    while (level > value_weight_pairs[idx + 1].second)
+                        idx++;
+                }
+            }
+
+            size_t l = idx;
+            size_t u = idx + 1 < size ? idx + 1 : idx;
+
+            Float64 xl = value_weight_pairs[l].second, xr = value_weight_pairs[u].second;
+            UnderlyingType yl = value_weight_pairs[l].first, yr = value_weight_pairs[u].first;
+
+            if (level < xl)
+                yr = yl;
+            if (level > xr)
+                yl = yr;
+
+            result[indices[level_index]] = static_cast<T>(interpolate(level, xl, xr, yl, yr));
+            ++level_index;
+        }
+    }
+
+    /// This ignores overflows that may arise during add, sub and mul operations and doesn't aim to provide exact results.
+    /// Since `the quantileInterpolatedWeighted` function itself relies mainly on approximation, this is fine.
+    UnderlyingType interpolate(Float64 level, Float64 xl, Float64 xr, UnderlyingType yl, UnderlyingType yr) const
```

Comment:
@evillique was taking a look at the ubsan failures.
Last time I got around ubsan errors for these add, sub, mult and div operations, by using functions like `commom::mulIgnoreOverflow` and `commonLLaddIgnoreOverflow` ... etc. It used to work correctly but it looks like they fail now.  I'm looking for suggestions about the best way to handle overflows in this interpolation that I'm doing.

Reply:
@evillique fixed to use `NO_SANITIZE_UNDEFINED` on the interpolate function which also simplifies the expression. I also observed similar patterns in a few places in `DateTime`, `Aggregate` functions where overflows and NaNs are expected as a result of arithmetic operations.

The [ubsan tests for AST fuzzer](https://s3.amazonaws.com/clickhouse-test-reports/38252/cda37e73bac1a3c36d2d5b2972ed66262f1be459/fuzzer_astfuzzerubsan//report.html) look good now.